File: app/src/test/java/com/mohsenoid/rickandmorty/view/character/list/CharacterListPresenterTest.java
package com.mohsenoid.rickandmorty.view.character.list;

import com.mohsenoid.rickandmorty.data.DataCallback;
import com.mohsenoid.rickandmorty.domain.Repository;
import com.mohsenoid.rickandmorty.model.CharacterModel;
import com.mohsenoid.rickandmorty.test.CharacterDataFactory;
import com.mohsenoid.rickandmorty.test.DataFactory;
import com.mohsenoid.rickandmorty.util.config.ConfigProvider;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.stubbing.Answer;

import java.util.ArrayList;
import java.util.List;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class CharacterListPresenterTest {

    @Mock
    Repository repository;

    @Mock
    ConfigProvider configProvider;

    @Mock
    CharacterListContract.View view;

    private CharacterListContract.Presenter presenter;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);

        presenter = new CharacterListPresenter(repository, configProvider);
        presenter.bind(view);
    }

    @Test
    public void testLoadCharactersCallsViewShowLoading() {
        // GIVEN
        stubConfigProviderIsOnline(true);

        // WHEN
        presenter.loadCharacters();

        // THEN
        verify(view, times(1)).showLoading();
    }

    @Test
    public void testLoadCharactersCallsViewShowOfflineMessageWhenIsNotOnline() {
        // GIVEN
        stubConfigProviderIsOnline(false);

        // WHEN
        presenter.loadCharacters();

        // THEN
        verify(view, times(1)).showOfflineMessage(false);
    }

    @Test
    public void testLoadCharactersCallsRepositoryQueryCharacters() {
        // GIVEN
        stubConfigProviderIsOnline(true);

        List<Integer> characterIds = new ArrayList<>();
        characterIds.add(DataFactory.randomInt());
        characterIds.add(DataFactory.randomInt());
        characterIds.add(DataFactory.randomInt());

        presenter.setCharacterIds(characterIds);

        // WHEN
        presenter.loadCharacters();

        // THEN
        verify(repository, times(1)).queryCharacters(eq(characterIds), any());
    }

    @Test
    public void testLoadCharactersCallsViewSetCharactersOnSuccess() {
        // GIVEN
        List<CharacterModel> characters = CharacterDataFactory.makeCharactersList(5);
        stubRepositoryQueryCharactersOnSuccess(characters);

        // WHEN
        presenter.loadCharacters();

        // THEN
        verify(view, times(1)).setCharacters(eq(characters));
    }

    @Test
    public void testLoadCharactersCallsHideLoadingOnSuccess() {
        // GIVEN
        List<CharacterModel> characters = CharacterDataFactory.makeCharactersList(5);
        stubRepositoryQueryCharactersOnSuccess(characters);

        // WHEN
        presenter.loadCharacters();

        // THEN
        verify(view, times(1)).hideLoading();
    }

    @Test
    public void testKillCharacterCallsRepositoryWhenCharacterIsAlive() {
        // GIVEN
        stubConfigProviderIsOnline(true);
        CharacterModel character = CharacterDataFactory.makeCharacter();
        character.setStatus(CharacterModel.ALIVE);

        // WHEN
        presenter.killCharacter(character);

        // THEN
        verify(repository, times(1)).killCharacter(eq(character.getId()), any());
    }

    @Test
    public void testLoadCharactersCallsShowMessageOnError() {
        // GIVEN
        String errorMessage = DataFactory.randomString();
        stubRepositoryQueryCharactersOnError(new Exception(errorMessage));

        // WHEN
        presenter.loadCharacters();

        // THEN
        verify(view, times(1)).showMessage(errorMessage);
    }

    @Test
    public void testLoadCharactersCallsHideLoadingOnError() {
        // GIVEN
        stubRepositoryQueryCharactersOnError(new Exception());

        // WHEN
        presenter.loadCharacters();

        // THEN
        verify(view, times(1)).hideLoading();
    }

    @Test
    public void testKillCharacterSkipCallingRepositoryWhenCharacterIsKilled() {
        // GIVEN
        stubConfigProviderIsOnline(true);
        CharacterModel character = CharacterDataFactory.makeCharacter();
        character.setIsKilledByUser(true);

        // WHEN
        presenter.killCharacter(character);

        // THEN
        verify(repository, times(0)).killCharacter(eq(character.getId()), any());
    }

    private void stubConfigProviderIsOnline(boolean isOnline) {
        when(configProvider.isOnline())
                .thenReturn(isOnline);
    }

    private void stubRepositoryQueryCharactersOnSuccess(List<CharacterModel> characters) {
        doAnswer((Answer<Void>) invocation -> {
            DataCallback<List<CharacterModel>> callback = invocation.getArgument(1);
            callback.onSuccess(characters);

            return null;
        }).when(repository).queryCharacters(any(), any());
    }

    private void stubRepositoryQueryCharactersOnError(Exception exception) {
        doAnswer((Answer<Void>) invocation -> {
            DataCallback<List<CharacterModel>> callback = invocation.getArgument(1);
            callback.onError(exception);

            return null;
        }).when(repository).queryCharacters(any(), any());
    }
}

File: app/src/test/kotlin/com/mohsenoid/rickandmorty/view/character/list/CharacterListPresenterTest.kt
package com.mohsenoid.rickandmorty.view.character.list

import com.mohsenoid.rickandmorty.data.exception.NoOfflineDataException
import com.mohsenoid.rickandmorty.data.mapper.CharacterDbMapper
import com.mohsenoid.rickandmorty.domain.Repository
import com.mohsenoid.rickandmorty.domain.model.ModelCharacter
import com.mohsenoid.rickandmorty.test.CharacterDataFactory
import com.mohsenoid.rickandmorty.test.DataFactory
import com.mohsenoid.rickandmorty.util.StatusProvider
import com.nhaarman.mockitokotlin2.any
import kotlinx.coroutines.runBlocking
import org.amshove.kluent.Verify
import org.amshove.kluent.VerifyNoFurtherInteractions
import org.amshove.kluent.VerifyNoInteractions
import org.amshove.kluent.When
import org.amshove.kluent.called
import org.amshove.kluent.calling
import org.amshove.kluent.itAnswers
import org.amshove.kluent.itReturns
import org.amshove.kluent.on
import org.amshove.kluent.that
import org.amshove.kluent.was
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.MockitoAnnotations

class CharacterListPresenterTest {

    @Mock
    lateinit var repository: Repository

    @Mock
    lateinit var statusProvider: StatusProvider

    @Mock
    lateinit var view: CharacterListContract.View

    private lateinit var presenter: CharacterListContract.Presenter

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        presenter = CharacterListPresenter(repository = repository, configProvider = statusProvider)
        presenter.bind(view)
    }

    @Test
    fun `test if loadCharacters calls view showLoading`() {
        runBlocking {
            // GIVEN

            // WHEN
            presenter.loadCharacters(DataFactory.randomIntList(5))

            // THEN
            Verify on view that view.showLoading() was called
        }
    }

    @Test
    fun `test if isOffline loadCharacters calls view showOfflineMessage`() {
        runBlocking {
            // GIVEN
            stubConfigProviderIsOnline(false)

            // WHEN
            presenter.loadCharacters(DataFactory.randomIntList(5))

            // THEN
            Verify on view that view.showOfflineMessage(isCritical = false) was called
        }
    }

    @Test
    fun `test loadCharacters calls repository queryCharactersByIds`() {
        runBlocking {
            // GIVEN
            val characterIds: List<Int> = listOf(
                DataFactory.randomInt(),
                DataFactory.randomInt(),
                DataFactory.randomInt()
            )

            // WHEN
            presenter.loadCharacters(characterIds)

            // THEN
            Verify on repository that repository.getCharactersByIds(characterIds = characterIds) was called
        }
    }

    @Test
    fun `test loadCharacters calls view setCharacters OnSuccess`() {
        runBlocking {
            // GIVEN
            val characters: List<ModelCharacter> =
                CharacterDataFactory.Entity.makeCharacters(count = 5)
            stubRepositoryGetCharactersByIdsOnSuccess(characters)

            // WHEN
            presenter.loadCharacters(DataFactory.randomIntList(5))

            // THEN
            Verify on view that view.setCharacters(characters = characters) was called
        }
    }

    @Test
    fun `test loadCharacters calls view showMessage OnError`() {
        runBlocking {
            // GIVEN
            val errorMessage = DataFactory.randomString()
            stubRepositoryGetCharactersByIdsOnError(Exception(errorMessage))

            // WHEN
            presenter.loadCharacters(DataFactory.randomIntList(5))

            // THEN
            Verify on view that view.showMessage(message = errorMessage) was called
        }
    }

    @Test
    fun `test loadCharacters calls view onNoOfflineData having NoOfflineDataException OnError`() {
        runBlocking {
            // GIVEN
            stubRepositoryGetCharactersByIdsOnError(NoOfflineDataException())

            // WHEN
            presenter.loadCharacters(DataFactory.randomIntList(5))

            // THEN
            Verify on view that view.onNoOfflineData() was called
        }
    }

    @Test
    fun `test loadCharacters calls view hideLoading OnSuccess`() {
        runBlocking {
            // GIVEN
            val characters: List<ModelCharacter> =
                CharacterDataFactory.Entity.makeCharacters(count = 5)
            stubRepositoryGetCharactersByIdsOnSuccess(characters)

            // WHEN
            presenter.loadCharacters(DataFactory.randomIntList(5))

            // THEN
            Verify on view that view.hideLoading() was called
        }
    }

    @Test
    fun `test loadCharacters calls view hideLoading OnError`() {
        runBlocking {
            // GIVEN
            stubRepositoryGetCharactersByIdsOnError(Exception())

            // WHEN
            presenter.loadCharacters(DataFactory.randomIntList(5))

            // THEN
            Verify on view that view.hideLoading() was called
        }
    }

    @Test
    fun `test killCharacter calls repository killCharacter when character isAlive`() {
        runBlocking {
            // GIVEN
            val characterId = DataFactory.randomInt()
            val character: ModelCharacter = CharacterDataFactory.Entity.makeCharacter(
                characterId = characterId,
                status = CharacterDbMapper.ALIVE,
                isAlive = true,
                isKilledByUser = false
            )

            // WHEN
            presenter.killCharacter(character)

            // THEN
            Verify on repository that repository.killCharacter(
                characterId = characterId
            ) was called
            VerifyNoFurtherInteractions on repository
        }
    }

    @Test
    fun `test killCharacter skips calling repository killCharacter when character isNotAlive`() {
        runBlocking {
            // GIVEN
            val characterId = DataFactory.randomInt()
            val character: ModelCharacter = CharacterDataFactory.Entity.makeCharacter(
                characterId = characterId,
                status = CharacterDbMapper.ALIVE,
                isAlive = true,
                isKilledByUser = true
            )

            // WHEN
            presenter.killCharacter(character)

            // THEN
            VerifyNoInteractions on repository
        }
    }

    private fun stubConfigProviderIsOnline(isOnline: Boolean) {
        When calling statusProvider.isOnline() itReturns isOnline
    }

    private suspend fun stubRepositoryGetCharactersByIdsOnSuccess(characters: List<ModelCharacter>) {
        When calling repository.getCharactersByIds(any()) itReturns characters
    }

    private suspend fun stubRepositoryGetCharactersByIdsOnError(exception: Exception) {
        When calling repository.getCharactersByIds(any()) itAnswers { throw exception }
    }
}
